File: 3_Stack/generateParentheses.go
package main

import (
	"fmt"
	"strings"
)

// generateParenthesis generates all combinations of well-formed parentheses for given n pairs.
func generateParenthesis(n int) []string {
	var stack []string // stack to track current combination
	var res []string   // result to store valid combinations

	// backtrack recursively generates combinations
	var backtrack func(int, int)
	backtrack = func(openN int, closedN int) {
		// If both open and closed parentheses count reach n, add the combination to result
		if openN == n && closedN == n && openN == closedN {
			res = append(res, strings.Join(stack, ""))
			return
		}
		// If there are less than n open parentheses, add one and backtrack
		if openN < n {
			stack = append(stack, "(")
			backtrack(openN+1, closedN)
			pop(&stack) // remove the last element added (backtrack)
		}
		// If there are less closed parentheses than open ones, add one and backtrack
		if closedN < openN {
			stack = append(stack, ")")
			backtrack(openN, closedN+1)
			pop(&stack) // remove the last element added (backtrack)
		}
	}
	backtrack(0, 0) // start backtrack with 0 open and closed parentheses
	return res
}

// pop removes the last element from the slice (similar to pop() in other languages)
func pop(list *[]string) {
	length := len(*list)
	*list = (*list)[:length-1]
}

// implementation
func main() {
	fmt.Println("Leetcode 22: Generate Parentheses\n")

	// Example 1
	example1 := 3
	fmt.Printf("Example 1: n = %d\n", example1)
	fmt.Println("Output:", generateParenthesis(example1))

	// Example 2
	example2 := 1
	fmt.Printf("Example 2: n = %d\n", example2)
	fmt.Println("Output:", generateParenthesis(example2))
}

File: 3_Stack/generate_parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> list[str]:
        # Initialize an empty stack to keep track of the current combination
        stack = []

        # Initialize an empty list to store the valid combinations
        res = []

        # Recursive function to generate valid combinations
        def backtrack(openN, closedN):
            # If the number of open and closed parentheses is equal to n, add the current combination to the result
            if openN == closedN == n:
                res.append("".join(stack))
                return

            # If the number of open parentheses is less than n, add an open parenthesis and recurse
            if openN < n:
                stack.append("(")
                backtrack(openN + 1, closedN)
                stack.pop()

            # If the number of closed parentheses is less than the number of open parentheses, add a closed parenthesis and recurse
            if closedN < openN:
                stack.append(")")
                backtrack(openN, closedN + 1)
                stack.pop()

        # Start the recursive function with initial counts of open and closed parentheses set to 0
        backtrack(0, 0)

        # Return the list of valid combinations
        return res
    
# implementation
answer = Solution()
print("Leetcode 22: Generate Parentheses\n")

# Example 1
example_1 = 3
print(f"Example 1: n = {example_1}")
print("Output:", answer.generateParenthesis(example_1))

# Example 2
example_2 = 1
print(f"Example 2: n = {example_2}")
print("Output:", answer.generateParenthesis(example_2))